client/src/main/java/io/pravega/client/stream/impl/PositionImpl.java

```diff
@@ -22,23 +22,28 @@
 import java.nio.ByteBuffer;
 import java.util.Collections;
 import java.util.HashMap;
+import java.util.List;
 import java.util.Map;
 import java.util.Map.Entry;
+import java.util.Objects;
 import java.util.Set;
 import java.util.stream.Collectors;
 import lombok.Builder;
-import lombok.EqualsAndHashCode;
 import lombok.SneakyThrows;
 
 import static io.pravega.common.io.serialization.RevisionDataOutput.COMPACT_LONG_MAX;
 
-@EqualsAndHashCode(callSuper = false)
 public class PositionImpl extends PositionInternal {
 
     private static final PositionSerializer SERIALIZER = new PositionSerializer();
-    private final Map<Segment, Long> ownedSegments;
+    private Map<Segment, Long> ownedSegments;
     private final Map<Segment, Range> segmentRanges;
 
+    // If this field is set, it means that we will need to apply the updates on the ownedSegments.
+    private transient List<Entry<Segment, Long>> updatesToSegmentOffsets;
```

Comment:
Given the constants you selected this is going to add ~30MB of memory overhead per reader in a user's application. I don't consider this tradeoff worthwhile. See above for how we can avoid this.

Reply:
You are right in the case that the copy of the list is necessary, but I think that the current approach of referencing it may be feasible and is the one that provides the actual read throughput speedup.

Reply:
No. That map will be ~30mb and there is one per EventStreamReader.